runtime/domain.c

```diff
@@ -1119,32 +1134,45 @@ static void* domain_thread_func(void* v)
 {
   struct domain_startup_params* p = v;
   struct domain_ml_values *ml_values = p->ml_values;
+  /* All paths, success or failure, must pass through the handshake
+   * with the parent (which is waiting for us). */
+  char *failure = NULL;
 #ifndef _WIN32
   sigset_t mask = *(p->mask);
   void * signal_stack;
 
   signal_stack = caml_init_signal_stack();
   if (signal_stack == NULL) {
-    caml_fatal_error("Failed to create domain: signal stack");
+    failure = "Failed to create domain: signal stack";
   }
 #endif
 
-  domain_create(caml_params->init_minor_heap_wsz);
+  domain_create(caml_params->init_minor_heap_wsz, p->parent->state);
+
+  if (!domain_self) {
+    failure = "Failed to create domain";
+  }
+
   /* this domain is now part of the STW participant set */
   p->newdom = domain_self;
 
   /* handshake with the parent domain */
-  caml_plat_lock(&p->parent->lock);
+  caml_plat_lock(&p->parent->interruptor.lock);
   if (domain_self) {
     p->status = Dom_started;
     p->unique_id = domain_self->interruptor.unique_id;
   } else {
     p->status = Dom_failed;
   }
-  caml_plat_broadcast(&p->parent->cond);
-  caml_plat_unlock(&p->parent->lock);
+  caml_plat_broadcast(&p->parent->interruptor.cond);
+  caml_plat_unlock(&p->parent->interruptor.lock);
   /* Cannot access p below here. */
 
+  /* Handshake completed: now we can fail! */
+  if (failure) {
+    caml_fatal_error("%s", failure);
+  }
```

Comment:
As `caml_fatal_error` will stop the program entirely, there is no point in awaiting the completion of the handshake.

Reply:
There has been discussion of surviving failures in `Domain.spawn`, with an exception. If we ever do that, this can become an exception-raising path.

Reply:
You might be referring to #12410. Then do not worry and let me handle it :) For now the smallest rebase conflict the simpler it is for me.